system/jlib/jstring.cpp

```diff
@@ -939,25 +939,73 @@ StringBuffer &replaceString(StringBuffer & result, size_t lenSource, const char
 {
     if (lenSource)
     {
-        size_t left = lenSource;
-        while (left >= lenOldStr)
+        if (lenOldStr)
         {
-            if (memcmp(source, oldStr, lenOldStr)==0)
+            // Scan for match on first character
+            size_t offset = 0;
+            size_t left = lenSource;
+            while (left >= lenOldStr)
             {
-                result.append(lenNewStr, newStr);
-                source += lenOldStr;
-                left -= lenOldStr;
+                if (source[offset] != oldStr[0])
+                {
+                    offset++;
+                    left--;
+                }
+                else
+                    break;
             }
-            else
+
+            // Reserve space when the new string is longer than the old string
+            unsigned steps = lenSource - lenOldStr - offset + 1;
+            unsigned maxResultLength = lenNewStr > lenOldStr ? lenSource + steps * (lenNewStr - lenOldStr) : lenSource;
```

Comment:
You can't guess how long the new string is likely to be, so it is simplest to assume it will match the source.  If there is a match and the replace string is shorter this will be slightly too long, if it is longer it will (possibly) reallocate.  The big advantage in preallocating is to avoid repeated copies as the buffer is built up.  This can become significant for large strings.  A single reallocation by comparison is minor.

Reply:
This attempts to calculate the max size if for every comparison there is a match. If the lenOldStr<=lenNewStr then it does assume it will match the source. I think this does preallocate enough space so there are no need for reallocations, but if we always assume the length matches the source then sometimes we won't preallocate enough.